### DISCUSSION TITLE
Type of variable different when instantiated in return

### DISCUSSION THREAD

**Discussion by haarisr** [microsoft/pyright#8569] [General]
```python3
import numpy as np
import numpy.typing as npt


def foo() -> npt.NDArray[np.float64]:
    return np.array([1.0, 2.0, 3.0], dtype=np.float32)

def bar() -> npt.NDArray[np.float64]:
    temp = np.array([1.0, 2.0, 3.0], dtype=np.float32)
    return temp
```

```
try.py:10:12 - error: Expression of type "NDArray[floating[_32Bit]]" is incompatible with return type "NDArray[float64]"
    "ndarray[Any, dtype[floating[_32Bit]]]" is incompatible with "ndarray[Any, dtype[float64]]"
      Type parameter "_DType_co@ndarray" is covariant, but "dtype[floating[_32Bit]]" is not a subtype of "dtype[float64]"
        "dtype[floating[_32Bit]]" is incompatible with "dtype[float64]"
          Type parameter "_DTypeScalar_co@dtype" is covariant, but "floating[_32Bit]" is not a subtype of "float64"
            "floating[_32Bit]" is incompatible with "floating[_64Bit]" (reportReturnType)
1 error, 0 warnings, 0 informations
```

I am not sure why `foo` does not return an error but `bar` does. I believe `foo` should also throw an error? I could not create an example with standard types, not sure if it is a `numpy` issue

**Comment by erictraut** [comment]
I'm not able to repro the behavior you're seeing. I'm guessing that you're using an older version of numpy. Please try upgrading to the latest version (`pip install numpy --upgrade`). With the latest version, I'm seeing the return statements in both `foo` and `bar` result in type violation errors.

The big difference between `foo` and `bar` is that the call to `np.array` is evaluated with normal inference rules in `bar` but with [bidirectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types) rules in `foo`. With bidirectional type inference, pyright takes into account the "expected type", which in this case is provided by the return type annotation.

**Reply by haarisr** [reply]
I still face the same error with numpy 2.0.1. 

![image](https://github.com/user-attachments/assets/ba35ad0b-c064-4981-8135-6dd75b88001f)

Seems like it evaluates to `NDArrray[Any]`?

**Reply by erictraut** [reply]
Perhaps you have an older type stub library for numpy installed in your python environment? You could confirm by creating a fresh venv and installing numpy.

**Reply by haarisr** [reply]
```bash
uv venv venv -p 3.12 && \
source venv/bin/activate && \
uv pip install pyright numpy --no-cache && \
uv pip show pyright && \
uv pip show numpy && \
pyright try.py && \
cat try.py

Using Python 3.12.4 interpreter at: /usr/bin/python3.12
Creating virtualenv at: venv
Activate with: source venv/bin/activate
Resolved 3 packages in 312ms
Prepared 3 packages in 1.80s
Installed 3 packages in 4ms
 + nodeenv==1.9.1
 + numpy==2.0.1
 + pyright==1.1.373
Name: pyright
Version: 1.1.373
Location: foo/venv/lib/python3.12/site-packages
Requires: nodeenv
Required-by:
Name: numpy
Version: 2.0.1
Location: foo/venv/lib/python3.12/site-packages
Requires:
Required-by:
0 errors, 0 warnings, 0 informations
import numpy as np
import numpy.typing as npt


def foo() -> npt.NDArray[np.float64]:
    return np.array([1.0, 2.0, 3.0], dtype=np.float32)
```
Still the same

**Reply by haarisr** [reply]
```python3

import numpy as np
import numpy.typing as npt


def foo() -> npt.NDArray[np.float64]:
    return (arr:=np.array([1.0, 2.0, 3.0], dtype=np.float32), reveal_type(arr))
```

I get the error with this but this might be similar to the bar example again. Without `reveal_type` there is no error

**Reply by erictraut** [reply]
Ah, my bad. I'm running the latest (prerelease) version of pyright. This turns out to be a bug that I fixed a couple of days ago (https://github.com/microsoft/pyright/issues/8519). It will be addressed in this week's release.

**Reply by haarisr** [reply]
Thanks for the help and the work on pyright!
